test/csuite/random_directio/main.c

```diff
@@ -0,0 +1,1264 @@
+
+		/*
+		 * Every 1000th record is large.
+		 */
+		if (id % 1000 == 0)
+			TEST_STREQ(large_arr[th], gotvalue,
+			    "main table large value");
+		else
+			TEST_STREQ(&keybuf[kvsize], gotvalue,
+			    "main table value");
+
+		/*
+		 * Check the reverse file, with key/value reversed.
+		 */
+		check_kv(rev, &keybuf[kvsize], keybuf, true);
+
+		check_schema(session, id, th, flags);
+
+		/* Bump thread number and id to the next expected key. */
+		th = (th + 1) % nth;
+		if (th == 0)
+			id++;
+	}
+	printf("scanned to %" PRIu64 "\n", id);
+
+	if (LF_ISSET(SCHEMA_ALL)) {
+		/*
+		 * Check metadata to see if there are any tables
+		 * present that shouldn't be there.
+		 */
+		testutil_check(session->open_cursor(session, "metadata:", NULL,
+		    NULL, &meta));
+		while ((ret = meta->next(meta)) != WT_NOTFOUND) {
+			testutil_check(ret);
+			meta->get_key(meta, &gotkey);
+			/*
+			 * Names involved in schema testing are off the form:
+			 *   table:Axxx-t
+			 *   table:Bxxx-t
+			 * xxx corresponds to the id inserted into the main
+			 * table when the table was created, and t corresponds
+			 * to the thread id that did this.
+			 */
+			if (WT_PREFIX_SKIP(gotkey, "table:") &&
+			    (*gotkey == 'A' || *gotkey == 'B')) {
+				gotid = (uint64_t)strtol(gotkey + 1, &p, 10);
+				testutil_assert(*p == '-');
+				th = (uint32_t)strtol(p + 1, &p, 10);
+				testutil_assert(*p == '\0');
+				/*
+				 * XXX
+				 * If table operations are transactional,
+				 * then there is more to do here.
+				 */
+			}
+		}
+		testutil_check(meta->close(meta));
+
+	}
+
+	testutil_check(cursor->close(cursor));
+	testutil_check(rev->close(rev));
+	testutil_check(session->close(session, NULL));
+	testutil_check(conn->close(conn, NULL));
+
+	for (th = 0; th < nth; th++)
+		free(large_arr[th]);
+	free(large_arr);
+	free(keybuf);
+	free(lastid);
+	return (true);
+}
+
+/*
+ * handler --
+ *	Child signal handler
+ */
+static void
+handler(int sig)
+{
+	pid_t pid;
+	int status, termsig;
+
+	WT_UNUSED(sig);
+	pid = waitpid(-1, &status, WNOHANG|WUNTRACED);
+	if (pid == 0)
+		return;		/* Nothing to wait for. */
+	if (WIFSTOPPED(status))
+		return;
+	if (WIFSIGNALED(status)) {
+		termsig = WTERMSIG(status);
+		if (termsig == SIGCONT || termsig == SIGSTOP)
+			return;
+		printf("Child got signal %d (status = %d, 0x%x)\n",
+		    termsig, status, (unsigned int)status);
+#ifdef WCOREDUMP
+		if (WCOREDUMP(status))
+			printf("Child process id=%d created core file\n", pid);
+#endif
+	}
+
+	/*
+	 * The core file will indicate why the child exited. Choose EINVAL here.
+	 */
+	testutil_die(EINVAL,
+	    "Child process %" PRIu64 " abnormally exited, status=%d (0x%x)",
+	    (uint64_t)pid, status, status);
+}
+
+/*
+ * has_direct_io --
+ *	Check for direct I/O support.
+ */
+static bool
+has_direct_io(void)
+{
+#ifdef O_DIRECT
+	return (true);
+#else
+	return (false);
+#endif
+}
+
+/*
+ * main --
+ *	Top level test.
+ */
+int
+main(int argc, char *argv[])
+{
```

Comment:
This is a large, comprehensive and complex test program. Do you want to consider breaking the code up into a few files for easier understanding? child.c, verify.c, main.c or others that seem suitable.

Reply:
@sueloverso, I think I'd do this as a followup.